[quant][graphmode][fx] Optimize cat

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54928 [quant][graphmode][fx][refactor] Quantize by Use of a Tensor instead of Tensor
* #54924 [quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat
* **#54813 [quant][graphmode][fx] Optimize cat**

Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

jerryzh168 committed:
[quant][graphmode][fx] Optimize cat

Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -491,6 +495,52 @@ def in_nodes(a: Argument, nodes: Set[Node]) -> bool:
+        modules: Dict[str, torch.nn.Module]):
+    observed_nodes: Set[Node] = set()
+    # activation_post_process for cat nodes
+    app_for_cat_nodes: Dict[Node, torch.nn.Module] = dict()
```
optional: maybe `actpp` instead of `app`?  `app` means so many things :)

jerryzh168 replied:
sure

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -491,6 +495,52 @@ def in_nodes(a: Argument, nodes: Set[Node]) -> bool:
+    # activation_post_process for cat nodes
+    app_for_cat_nodes: Dict[Node, torch.nn.Module] = dict()
+
+    # we'll set the activation_post_process for all tensor inputs of cat and output
+    # of cat to be the same (the activation_post_process for the first Tensor
+    # input in the list, for example, if we have:
+    # x = torch.cat([x1, x2, x3], ...)
```
is it

```
x1 -> obs1 -> cat -> obs1 -> ...
              /
      x2 -> obs1
```

or 

```
x1 -> list -> obs1 -> cat -> obs1 -> ...
        /
     x2
```

or something else?  Perhaps we can include the ascii of what's going on in the docblock?

jerryzh168 replied:
the first one, we don't have observers for list I think

vkuzo replied:
makes sense.  Could we add the graph to the docblock?  It would be really helpful for code readers.  

Seems like sharing observers for inputs of different sizes would bias the qparams towards the tensors with smaller sizes.  A workaround could be do weigh the contribution of different components of cat by the number of elements, but probably not worth it.  This PR is already much better compared to what we were doing before.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Optimize cat"


Summary:
Previously we have a cat that takes a list of Tensors with different qparams and dequantize them
cacatenate them and requantize with the output qparams. This adds some unnecessary overhead in dequantizing
and quantizing Tensors.

This PR adds an optimization for cat operator, we'll make sure inputs and output of cat
uses same observer/fake_quant and produce a cat that does not do rescaling.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27408377](https://our.internmc.facebook.com/intern/diff/D27408377)

[ghstack-poisoned]
